### PR TITLE
feat!: raise error instead of silently returning on interactions that do not support deferring

### PR DESCRIPTION
**Pull request by onerandomusername**
## Summary

small update of typehints and documentation of defer handling,
and make it a bit more clear how it works.

code is directly taken from #474, but was removed to lower feature creep.

## Checklist

- [x] If code changes were made, then they have been tested
    - [x] I have updated the documentation to reflect the changes
    - [x] I have formatted the code properly by running `task lint`
    - [x] I have type-checked the code by running `task pyright`
- [ ] This PR fixes an issue
- [x] This PR adds something new (e.g. new method or parameters)
- [x] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

**Comment by Chromosomologist** [comment]
LGTM!
I don't see anything weird in the code, the docstring makes sense to me and the overloads seem to match what is described in said docstring.

File: disnake/interactions/base.py [hunk 1]
```diff
@@ -689,23 +733,35 @@ async def defer(self, *, ephemeral: bool = False, with_message: bool = False) ->
...
+
+        if parent.type in (InteractionType.application_command, InteractionType.modal_submit):
+            defer_type = InteractionResponseType.deferred_channel_message
         elif parent.type is InteractionType.component:
-            defer_type = InteractionResponseType.deferred_message_update.value
+            if with_message:
+                defer_type = InteractionResponseType.deferred_channel_message
+            else:
+                defer_type = InteractionResponseType.deferred_message_update
+
```
Comment (shiftinv) [inline disnake/interactions/base.py:750]: could raise here instead of silently dropping the call, if none of the types match

Comment (onerandomusername) [reply disnake/interactions/base.py:750]: Fair, I initially didn't want to make it a breaking change but it *really* should raise instead.

File: disnake/interactions/base.py [hunk 2]
```diff
@@ -659,7 +659,27 @@ def is_done(self) -> bool:
...
+        with_message: bool = ...,
+        ephemeral: bool = ...,
+    ) -> None:
+        ...
+
+    @overload
+    async def defer(
+        self,
+    ) -> None:
+        ...
```
Comment (shiftinv) [inline disnake/interactions/base.py:675]: These overloads don't seem necessary

File: disnake/interactions/base.py [hunk 3]
```diff
@@ -669,16 +689,14 @@ async def defer(self, *, ephemeral: bool = False, with_message: bool = False) ->
 
         Parameters
```
Comment (shiftinv) [inline disnake/interactions/base.py:690]: Default values should be noted here, at least for `ephemeral`

Comment (onerandomusername) [reply disnake/interactions/base.py:690]: The default for ephemeral has the same behavior as other methods. Not sure why it needs to be documented here.

File: disnake/interactions/base.py (hunk 2 above)
Comment (onerandomusername) [reply disnake/interactions/base.py]: https://github.com/DisnakeDev/disnake/pull/505/commits/c94008b6420d366c613ae4dc7d6ac0530633a74c

**Review by shiftinv — APPROVED** [review]

**Pull request merged**